### PR TITLE
Feature/eth sign

### PR DESCRIPTION
**Pull request by kazunetakeda25**
Added `eth_sign` method

### CODE REVIEW

File: src/Lachain.Core/RPC/HTTP/Web3/AccountServiceWeb3.cs [hunk 1]
```diff
@@ -170,10 +179,32 @@ public JArray GetAccounts()
...
+            {
+                throw new Exception("address is invalid");
+            }
+
+            var keyPair = _privateWallet.EcdsaKeyPair;
+
+            Logger.LogInformation($"Keys: {keyPair.PublicKey.GetAddress().ToHex()}");
+
+            var messageHashBytes = messageBytes.KeccakBytes();
+            var signed = crypto.SignHashed(messageHashBytes, keyPair.PrivateKey.Encode());
```
Comment (sgladkov) [inline src/Lachain.Core/RPC/HTTP/Web3/AccountServiceWeb3.cs:205]: As I said,  don't just sign received message,  add some constant prefix to the message before signing,  for example sign smth like "LACHAIN" + messageLength + message

File: src/Lachain.Core/RPC/HTTP/Web3/AccountServiceWeb3.cs [hunk 2]
```diff
@@ -170,10 +179,32 @@ public JArray GetAccounts()
...
+
+            byte[]? messageBytes = message.HexToBytes();
+
+            if (_privateWallet.IsLocked())
+            {
+                throw new Exception("wallet is locked");
+            }
+
+            var addressUInt160 = (address!).HexToUInt160();
+            if (addressUInt160.Equals(_privateWallet.EcdsaKeyPair.PublicKey.GetAddress()))
```
Comment (tbssajal) [inline src/Lachain.Core/RPC/HTTP/Web3/AccountServiceWeb3.cs:195]: I think it should be NotEquals

Comment (sgladkov) [reply src/Lachain.Core/RPC/HTTP/Web3/AccountServiceWeb3.cs:195]: Good catch.)

Comment (kazunetakeda25) [reply src/Lachain.Core/RPC/HTTP/Web3/AccountServiceWeb3.cs:195]: Hi @tbssajal Yes right :)

File: src/Lachain.Core/RPC/HTTP/Web3/AccountServiceWeb3.cs [hunk 3]
```diff
@@ -170,10 +181,32 @@ public JArray GetAccounts()
...
-            // TODO: implement message signing
-            //var addressUint160 = address.HexToUInt160();
-            return Web3DataFormatUtils.Web3Data("".HexToBytes());
-            //throw new ApplicationException("Not implemented yet");
+            if (address is null)
+            {
+                throw new ArgumentException("address should not be null");
+            }
+
+            byte[]? messageBytes = (_messagePrefix + message.Length + message).HexToBytes();
```
Comment (sgladkov) [inline src/Lachain.Core/RPC/HTTP/Web3/AccountServiceWeb3.cs:189]: It will not work,  Prefix is not correct hex string,  so HexToBytes will fail

**Review by sgladkov — CHANGES_REQUESTED** [review]

File: src/Lachain.Core/RPC/HTTP/Web3/AccountServiceWeb3.cs [hunk 4]
```diff
@@ -170,10 +183,32 @@ public JArray GetAccounts()
...
-            // TODO: implement message signing
-            //var addressUint160 = address.HexToUInt160();
-            return Web3DataFormatUtils.Web3Data("".HexToBytes());
-            //throw new ApplicationException("Not implemented yet");
+            if (address is null)
+            {
+                throw new ArgumentException("address should not be null");
+            }
+
+            byte[]? messageBytes = Encoding.ASCII.GetBytes(_messagePrefix + message.Length + message);
```
Comment (sgladkov) [inline src/Lachain.Core/RPC/HTTP/Web3/AccountServiceWeb3.cs:191]: No,  it is difficult to verify this signature on the client side, Use `Encoding.ASCII.GetBytes` for conversion of the  `_messagePrefix + message.Length` and `HexToBytes` for conversion of the `message`,  after that concatenate 2 ByteArrays

**Review by sgladkov — CHANGES_REQUESTED** [review]

**Review by sgladkov — APPROVED** [review]

File: src/Lachain.Core/RPC/HTTP/Web3/AccountServiceWeb3.cs [hunk 5]
```diff
@@ -12,30 +12,39 @@
...
         private readonly IStateManager _stateManager;
         private readonly ISnapshotIndexRepository _snapshotIndexer;
         private readonly IContractRegisterer _contractRegisterer;
         private readonly ISystemContractReader _systemContractReader;
         private readonly ITransactionPool _transactionPool;
+        private readonly IPrivateWallet _privateWallet;
 
-        
-
+        private readonly DefaultCrypto crypto = new DefaultCrypto();
```
Comment (tbssajal) [inline src/Lachain.Core/RPC/HTTP/Web3/AccountServiceWeb3.cs:37]: I think it should be something like this: https://github.com/LATOKEN/lachain/blob/dev/src/Lachain.Core/Blockchain/Operations/TransactionSigner.cs#L12

**Review by tbssajal — DISMISSED** [review]

File: src/Lachain.Core/RPC/HTTP/Web3/AccountServiceWeb3.cs (hunk 5 above)
Comment (kazunetakeda25) [reply src/Lachain.Core/RPC/HTTP/Web3/AccountServiceWeb3.cs]: fixed, thanks

**Review by sgladkov — DISMISSED** [review]
For the future: don't add format changes to functional pull request please,  it makes review more difficult.

**Review by HMAshiqulIslam — CHANGES_REQUESTED** [review]
Adding an extra space after "if" and "foreach" makes it inconsistent throughout the codebase. We already follow a convention and the whole codebase is based on that. We should follow the existing conventions.
=> Got it. fixed

**Review by sgladkov — APPROVED** [review]

**Review by HMAshiqulIslam — COMMENTED** [review]
Can you please add some unit test / testing procedure to test this api call ?

**Pull request closed without merging**